lib/auth/auth_lib.c

```diff
@@ -0,0 +1,342 @@
+ *
+ * @param arg1  Pointer to struct auth_thread_params.
+ * @param arg2  Unused
+ * @param arg3  Unused
+ */
+static void auth_thrd_entry(void *arg1, void *arg2, void *arg3)
+{
+	int ret;
+	struct auth_thread_params *thrd_params = (struct auth_thread_params *)arg1;
+
+	while (true) {
+
+		ret = k_sem_take(thrd_params->thrd_sem, K_FOREVER);
+
+		if (ret) {
+			LOG_ERR("Failed to get semaphore for auth thread, err: %d", ret);
+			LOG_ERR("Auth thread terminating, instance id: %d.",
+				thrd_params->auth_conn->instance);
+			return;
+		}
+
+		/* call auth thread function */
+		thrd_params->auth_conn->auth_func(thrd_params->auth_conn);
+	}
+}
+
+
+
+/* ========================= external API ============================ */
+
+
+/**
+ * @see auth_lib.h
+ */
+int auth_lib_init(struct authenticate_conn *auth_conn, enum auth_instance_id instance,
+		  auth_status_cb_t status_func, void *context, struct auth_optional_param *opt_params,
+		  enum auth_flags auth_flags)
+{
+	int err = 0;
+
+	/* check input params */
+	if (status_func == NULL) {
+		LOG_ERR("Error, status function is NULL.");
+		return AUTH_ERROR_INVALID_PARAM;
+	}
+
+	/* check auth flags */
+	if (!auth_lib_checkflags(auth_flags)) {
+		LOG_ERR("Invalid auth flags.");
+		return AUTH_ERROR_INVALID_PARAM;
+	}
+
+	/* init the struct to zero */
+	memset(auth_conn, 0, sizeof(struct authenticate_conn));
+
+	/* setup the status callback */
+	auth_conn->status_cb = status_func;
+	auth_conn->callback_context = context;
+
+	auth_conn->cancel_auth = false;
+	auth_conn->instance = instance;
+
+	/* init the work item used to post authentication status */
+	k_work_init(&auth_conn->auth_status_work, auth_lib_status_work);
+
+	auth_conn->is_client = (auth_flags & AUTH_CONN_CLIENT) ? true : false;
+
+#if defined(CONFIG_AUTH_DTLS)
+
+	if (auth_flags & AUTH_CONN_DTLS_AUTH_METHOD) {
+		/* Set the DTLS authentication thread */
+		auth_conn->auth_func = auth_dtls_thead;
+		{
+			if (opt_params == NULL || opt_params->param_id != AUTH_DTLS_PARAM) {
+				LOG_ERR("Missing certificates for TLS/DTLS authentication.");
+				return AUTH_ERROR_INVALID_PARAM;
+			}
+
+			struct auth_dtls_certs *certs = &opt_params->param_body.dtls_certs;
+
+			// init TLS layer
+			err = auth_init_dtls_method(auth_conn, certs);
+		}
+
+		if (err) {
+			LOG_ERR("Failed to initialize MBed TLS, err: %d", err);
+			return err;
+		}
+	}
+#endif
+
+#if defined(CONFIG_AUTH_CHALLENGE_RESPONSE)
+
+	if (auth_flags & AUTH_CONN_CHALLENGE_AUTH_METHOD) {
+		/* Set the Challenge-Response authentication thread */
+		auth_conn->auth_func = auth_chalresp_thread;
+
+		if ((opt_params != NULL) && (opt_params->param_id == AUTH_CHALRESP_PARAM)) {
+
+			struct auth_challenge_resp *chal_resp = &opt_params->param_body.chal_resp;
+
+			err = auth_init_chalresp_method(auth_conn, chal_resp);
+
+			if (err) {
+				LOG_ERR("Failed to set Challege-Response param, err: %d", err);
+				return err;
+			}
+		}
+	}
+#endif
```

Comment:
these `ifdef` don't scale well, if you have N backends this code will get very complicate. You should have an API that is implemented by the backend and here you should just call the init callback.

Reply:
Makes sense, the common code should not have to know about the individual auth method details. Fixed, moved backend specific init code to the backend init function.